Related Products: avoids querying results on each page load

cpapazoglou opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR improves the performance of related products handling in WooCommerce by:

1. Optimizing related products caching to query only once per day instead of on every page load
2. Moving related product transient deletion to an async process to prevent blocking operations

Closes https://github.com/woocommerce/woocommerce/issues/30349

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install Query Monitor
2. Edit the single product templated to show related products by category
3. Create three products in the same category ( master product, 2x related products)
4. Load master product page 
5. The related products should be shown
6. See list of 'UPDATE' queries from QM console
7. Reloading the page should not do consequent `UPDATEs` ( don't query related products on each page load )
8. Edit one related product and remove it from the category
9. Reload master product page
10. The related product should not be shown ( transient cleared )

|Before | After|
|-------|------|
|![CleanShot 2025-02-25 at 20 06 09](https://github.com/user-attachments/assets/91e7ad5f-d925-4cd1-85bb-5e84f2ec6ccf)|![CleanShot 2025-02-25 at 20 05 41](https://github.com/user-attachments/assets/797ee550-83b1-4932-b739-00de1af59c8a)|

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [x] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Query related products once per day when loading product details page

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Commit:
Related Products: avoids querying when results<limit

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Related Products: properly bust the transients.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

cpapazoglou commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -1834,3 +1835,34 @@ function wc_product_attach_featured_image( $attachment_id, $product = null, $sav
+ * @param int $product_id Product ID.
+ */
+function wc_delete_related_product_transients( $product_id ) {
+	global $wpdb;
```
I was looking at returning early only for new products when there isn't a category and tag set but by default all products get the `uncategorised` category 🤦 .

Also, I would like to bail early if there is no change in categories or tags when updating a product but couldn't find a way to identify changes.

Aljullu replied:
Those were my first thoughts when reviewing this PR, but I understand we don't have access to the previous categories/tags and because of that we have no way to know which products need to be updated, right?

But if that's the case, do we need to pass `$product_id` into this function? :thinking:

cpapazoglou replied:
Good catch! I had a call with Tom and decided to delete the transients only for the affected products ( old related products and new related products ) so function signature will change anyway. Thanks for the review!

cpapazoglou commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -1834,3 +1835,34 @@ function wc_product_attach_featured_image( $attachment_id, $product = null, $sav
 add_action( 'add_attachment', 'wc_product_attach_featured_image' );
+
+/**
+ * Delete related product transients when a product is saved.
```
Reasoning behind this nuclear approach https://github.com/woocommerce/woocommerce/issues/30349#issuecomment-2683037010

Commit:
Related Products: simplifies the call of the function.

Aljullu commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -1834,3 +1835,30 @@ function wc_product_attach_featured_image( $attachment_id, $product = null, $sav
+function wc_delete_related_product_transients( $product_id ) {
+	global $wpdb;
+
+	// Delete all related product timeout transients
+	$wpdb->query( 
+		$wpdb->prepare( 
+			"DELETE FROM {$wpdb->options} WHERE option_name LIKE %s",
+			$wpdb->esc_like( '_transient_timeout_wc_related_' ) . '%'
+		)
+	);
+
+	// Also delete the actual transient entries
+	$wpdb->query( 
+		$wpdb->prepare( 
+			"DELETE FROM {$wpdb->options} WHERE option_name LIKE %s",
+			$wpdb->esc_like( '_transient_wc_related_' ) . '%'
+		)
+	);
```
Not a big deal, but wouldn't it be better to do this in one query instead of two? Or is there a reason to prefer doing two queries?

```suggestion
	// Delete both timeout and actual transient entries in one query
	$wpdb->query( 
	    $wpdb->prepare( 
	        "DELETE FROM {$wpdb->options} WHERE option_name LIKE %s OR option_name LIKE %s",
	        $wpdb->esc_like( '_transient_timeout_wc_related_' ) . '%',
	        $wpdb->esc_like( '_transient_wc_related_' ) . '%'
	    )
	);
```

cpapazoglou replied:
Good idea, will do!

Aljullu reviewed (commented):
Thanks for working on this, @cpapazoglou! This is not my area of expertise, but I tried to do a review anyway.

One concern I have is that we are no longer calling [`delete_transient()`](https://developer.wordpress.org/reference/functions/delete_transient/) to delete those transients, so there are some actions that won't be triggered (ie: [`delete_transient_wc_related_ID`](https://developer.wordpress.org/reference/hooks/delete_transient_transient/)). I guess we could trigger the actions manually, but we would need to know which products are being updated.

I took a quick look and didn't find any extensions listening to those actions, but wondering about your thoughts on this. :thinking:

Commit:
Related Products: clear only related transients async

Commit:
Related products: fixes async call, deletes current transient

cpapazoglou commented:
- Changed the function to clears only affected related products
- Moved the logic to the scheduler to avoid performance issues from calling `wc_get_related_products( $post_id, 1000 );`

Noticed that `[wc_delete_product_transients](https://github.com/woocommerce/woocommerce/blob/91d6172e4bac8159492972b477abde2bed28a62a/plugins/woocommerce/includes/wc-product-functions.php#L127)` is called twice when saving a product. I will investigate why.

![CleanShot 2025-02-26 at 17 18 32](https://github.com/user-attachments/assets/6e36d2a8-c1ee-42eb-b3ce-81c304f34806)

cpapazoglou commented:
Alright the duplicate save product call comes from `class-wc-meta-box-product-data.php:437` and `class-wc-meta-box-product-images.php:101` which seem unrelated to these changes and should be investigated in a separate issue.

Stack Traces
```
2025-02-26T15:45:22+00:00 Debug Product save() called. Stack trace: 
WC_Product::save in /woocommerce/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-data.php on line 437
WC_Meta_Box_Product_Data::save in /wp-includes/class-wp-hook.php on line 324
WP_Hook::apply_filters in /wp-includes/class-wp-hook.php on line 348
WP_Hook::do_action in /wp-includes/plugin.php on line 517
do_action in /woocommerce/plugins/woocommerce/includes/admin/class-wc-admin-meta-boxes.php on line 273
WC_Admin_Meta_Boxes::save_meta_boxes in /wp-includes/class-wp-hook.php on line 326
WP_Hook::apply_filters in /wp-includes/class-wp-hook.php on line 348
WP_Hook::do_action in /wp-includes/plugin.php on line 517
do_action in /wp-includes/post.php on line 5060
wp_insert_post in /wp-includes/post.php on line 5162
wp_update_post in /wp-admin/includes/post.php on line 453
edit_post in /wp-admin/post.php on line 227

2025-02-26T15:45:22+00:00 Debug Product save() called. Stack trace: 
WC_Product::save in /woocommerce/plugins/woocommerce/includes/admin/meta-boxes/	
WC_Meta_Box_Product_Images::save in /wp-includes/class-wp-hook.php on line 324
WP_Hook::apply_filters in /wp-includes/class-wp-hook.php on line 348
WP_Hook::do_action in /wp-includes/plugin.php on line 517
do_action in /woocommerce/plugins/woocommerce/includes/admin/class-wc-admin-meta-boxes.php on line 273
WC_Admin_Meta_Boxes::save_meta_boxes in /wp-includes/class-wp-hook.php on line 326
WP_Hook::apply_filters in /wp-includes/class-wp-hook.php on line 348
WP_Hook::do_action in /wp-includes/plugin.php on line 517
do_action in /wp-includes/post.php on line 5060
wp_insert_post in /wp-includes/post.php on line 5162
wp_update_post in /wp-admin/includes/post.php on line 453
edit_post in /wp-admin/post.php on line 227
```

cpapazoglou committed:
Merge branch 'trunk' into fix/related-products-db-update-query

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

tjcafferkey committed:
Linting

tjcafferkey commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -158,6 +173,67 @@ function wc_delete_product_transients( $post_id = 0 ) {
+			"DELETE FROM {$wpdb->options} WHERE option_name IN ($placeholders)",
+			$transient_names
+		)
+	);
```
Not all transients are stored in the database. We can implement a check for this like the below (haven't verified it though)

```suggestion
	// Delete specific timeout and transient entries.
	if ( wp_using_ext_object_cache() ) {
		// When using external object cache, we need to delete each transient individually
		foreach ( $related_product_ids as $id ) {
			delete_transient( 'wc_related_' . $id );
		}
	} else {
		// For database storage, we can use a single query for better performance
		$wpdb->query(
			$wpdb->prepare(
				"DELETE FROM {$wpdb->options} WHERE option_name IN ($placeholders)",
				$transient_names
			)
		);
	}
```

tjcafferkey commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -158,6 +173,67 @@ function wc_delete_product_transients( $post_id = 0 ) {
+ * Asynchronously delete related product transients when a product is changed.
+ * This is necessary because changing one product might affect many related products.
+ *
+ * @since 9.8.1
```
This should be `9.8.0` since feature freeze for the release is on Moday, beyond that it would be `9.9.0`

tjcafferkey commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -142,14 +143,28 @@ function wc_delete_product_transients( $post_id = 0 ) {
+		} else {
+			// Schedule the async deletion of related product transients.
+			// This should run async cause it also fetches the
+			// related products of the current product.
```
I think expanding on this comment would be beneficial. Something to the effect of "_Scheduled async because we can't be sure how many related product transients we're going to query_"

tjcafferkey commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -1043,8 +1117,9 @@ function wc_get_related_products( $product_id, $limit = 5, $exclude_ids = array(
-	// We want to query related posts if they are not cached, or we don't have enough.
-	if ( false === $related_posts || count( $related_posts ) < $limit ) {
+	// Query related posts if they are not cached
+	// Should happen only once per day due to transient expiriation set to 1 DAY_IN_SECONDS, to avoid performance issues
```
Lint 
```suggestion
	// Should happen only once per day due to transient expiriation set to 1 DAY_IN_SECONDS, to avoid performance issues.
```

tjcafferkey commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -456,12 +532,10 @@ function wc_get_formatted_variation( $variation, $flat = false, $include_names =
-			} else {
-				if ( $flat ) {
+			} elseif ( $flat ) {
 					$variation_list[] = rawurldecode( $value );
```
```suggestion
				$variation_list[] = rawurldecode( $value );
```

tjcafferkey commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -158,6 +173,67 @@ function wc_delete_product_transients( $post_id = 0 ) {
+	// Delete specific timeout and transient entries in a single query.
+	$wpdb->query(
+		$wpdb->prepare(
+			"DELETE FROM {$wpdb->options} WHERE option_name IN ($placeholders)",
```
I think this should be `({$placeholders})` for interoperability. But also I don't think the WP standards like using variables for this. You might have to inline it.

tjcafferkey commented on plugins/woocommerce/includes/wc-product-functions.php:
```diff
@@ -158,6 +173,67 @@ function wc_delete_product_transients( $post_id = 0 ) {
+	$old_transient        = get_transient( $transient_name );
+	$old_related_products = array();
+
+	if ( is_array( $old_transient ) ) {
```
Lets also check it's not an empty array

```suggestion
	if ( is_array( $old_transient ) && ! empty( $old_transient ) ) {
```

tjcafferkey reviewed (changes requested):
Thanks for this @cpapazoglou. Overall I like the approach, and understand why we're scheduling an async action. I've just proposed some changes let me know what you think 😄

tjcafferkey commented:
It would also be great if we could include a description of the change in this PR 😃

Commit:
Related Products: improve transient deletion and handling

cpapazoglou commented:
> Thanks for this @cpapazoglou. Overall I like the approach, and understand why we're scheduling an async action. I've just proposed some changes let me know what you think 😄

Thanks Tom, I have addressed your feedback in https://github.com/woocommerce/woocommerce/pull/55843/commits/a2fb4180827329f15c2e0de599e5b8d35f592e6f

Commit:
Linting

tjcafferkey commented:
@cpapazoglou when I adding/removing a related product from the category and refresh the master product, the related product does not disappear/reappear as expected. I even gave it a few minutes thinking the action had not run but to not avail.

But, when I went to a different product and then _returned_ to my master product the related product was present/absent as expected.

Is there something triggering this only when I am visiting a new product page, rather than refreshing my current (master) product?

cpapazoglou commented:
> @cpapazoglou when I adding/removing a related product from the category and refresh the master product, the related product does not disappear/reappear as expected. I even gave it a few minutes thinking the action had not run but to not avail.
> 
> But, when I went to a different product and then _returned_ to my master product the related product was present/absent as expected.
> 
> Is there something triggering this only when I am visiting a new product page, rather than refreshing my current (master) product?

I believe this is a scheduling / timing thing. Can you monitor http://localhost:10003/wp-admin/admin.php?page=wc-status&tab=action-scheduler&orderby=schedule&order=desc whether the action has run?

Sometimes it's instant and other times can take up to 60secs in my testing.

Commit:
Merge remote-tracking branch 'origin/trunk' into fix/related-products-db-update-query

Commit:
Uses the newly created _wc_delete_transients

Commit:
Updated docblock

tjcafferkey committed:
Merge branch 'trunk' into fix/related-products-db-update-query

tjcafferkey reviewed (approved):
Thanks @cpapazoglou this works as expected and fixes the original issue. There are still CI checks to pass but I am pre-approving. Once they have all passed we can go ahead and merge 🙌🏻

Commit:
Fixes failing test
